Proposal that fixes #13

Kertudex opened the pull request:
Hello, 

I tried to fix #13 with this PR. There is no possibility to know in advance where the ElasticaProxy Request is needed. I decided to add a parameter in the finder service definition to decide where to enable the finder. For the other actions (batchAction, exportAction), the default ProxyQuery is used.

I tested using behat and the features from the sandbox, it worked.

I'm quite new to github so sorry if some things are missing.

greg0ire commented:
Ok, thanks, let's review the code!

Kertudex commented:
The idea was to decide if the query is smart and if elastic search is needed in the same location.

greg0ire commented:
> The idea was to decide if the query is smart and if elastic search is needed in the same location.

Ok, that's good. With git, when you do the following
- commit A with mistake
- commit B
- commit that fixes mistake in commit A

and your commit are not merged yet, you can rewrite your history with `git rebase`, and end up with something easier to review, like this: 
- commit A
- commit B

greg0ire commented:
> I'm quite new to github so sorry if some things are missing.

No worries, and thank you for your contribution, I think it is quite good. However, what about BC ? Don't people need to change their code if they use your changes ?

Kertudex commented:
ES is used almost for the list action, that's why the default parameter of the "actions" config item is array("list").
I don't think it will break the compatibility. The only case would be to have its own search action without overwriting the listAction method in the CRUD controller (very strange). Anyway, currently, the batch action and the exports is not working at all. Thank you for the git lesson. I'll fix the commit and the PR quickly.

greg0ire commented:
I think you should try to end up with two commits only : 
- one commit for your feature
- one commit for the cs for problems you did not introduce yourself.

greg0ire commented:
ES is used almost for the list action, that's why the default parameter of the "actions" config item is array("list").
I don't think it will break the compatibility. The only case would be to have its own search action without overwriting the listAction method in the CRUD controller (very strange).

> ok, I think you're right

Anyway, currently, the batch action and the exports is not working at all. Thank you for the git lesson. I'll fix the commit and the PR quickly. 

Thanks! I think [this game](http://pcottle.github.io/learnGitBranching/) can help you to understand rebase if you need to.

OskarStark commented on Builder/ElasticaDatagridBuilder.php:
```diff
@@ -129,6 +129,17 @@ public function getBaseDatagrid(AdminInterface $admin, array $values = array())
     public function isSmart(AdminInterface $admin, array $values = array())
     {
+
+        //first : validate if esatica is asked in the configuration for this action
```
esatica -> elastica

OskarStark commented on Builder/ElasticaDatagridBuilder.php:
```diff
@@ -129,6 +129,17 @@ public function getBaseDatagrid(AdminInterface $admin, array $values = array())
+        $currentAction = explode('::', $fullCurrentAction);
+        //remove Action from listAction = list
+        $currentAction = substr($currentAction[1], 0, strlen($currentAction[1]) - 6);
+        //in case of batch|export action, no need to elastic search
```
elastics search -> elasticsearch

no space

OskarStark commented on README.md:
```diff
@@ -24,6 +24,7 @@ sonata_admin_search:
     admin_finder_services:
         my_admin.id: # Admin service id
             finder: Id for a FOS finder service that should be used # Finder service
+            actions : [list] #[Optional] actions where elastic search has to be enabled
```
can you please put this comment at the `Configuration.php`, too.

use `->info('my funky help')`

OskarStark commented on Tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -32,12 +32,14 @@ public function testProcessing()
                 array('admin_finder_services' => array(
                     'my_admin' => array(
                         'finder' => 42,
+
```
why?

OskarStark commented:
great work @Kertudex :+1:

Kertudex commented:
I tried to do it better but git doesn't like me :)
Is that better with the commits I did today ?

greg0ire commented:
I'll be able to review that tonight. I think it is better but I am not sure because I'm using my phone. If it isn't, I will do it for you.

greg0ire commented:
Ok, so at that point, there are way too many commits. You could play with `rebase`, but since the goal is to have only two commits, it would be even simpler to just forget the commits (but not the changes), and then commit the changes again. Here is what you should do to achieve that : 
1. Forget the commits : `git reset --mixed origin/master`. `mixed` is the default, so you do not really need it. There are three trees in git : the working directory, the staging area, and the repository. With `mixed`, you will reset the repository, the staging area, but not the working directory. `soft would reset only the repo, and`hard` would reset all three commits.
2. Add only the relevant changes to the staging area, using `git add --patch`. The `patch` option has two advantages :
   - It forces you to proofread what you are going to commit.
   - It enables you to commit part of a file.
     The commit message for the non-cs commit should contain `fixes #13` so that #13 is automatically closed when I merge. Once the first commit is done, use `git add -p` (abbreviation for patch) again to commit the cs changes.
3. Use `git push --force` to push changes and overwrite what you already pushed.

If you don't want to try, just say so, and I will do all this for you, but I think it is important you learn how to do that.

Kertudex committed:
fix #13
Added a parameter to choose when to use elastic search
The idea is to avoid the ElasticaProxyQuery usage in batch actions and
export.
We can now specify where to use elastic search in the configuration.
The default option uses it for the listAction only
the unit tests have been fixed accordingly

Kertudex committed:
fix cs

Kertudex commented:
I think the previous try was good (only 2 commits in my local branch) but i didn't know the --force option for the git push. Thank you.

greg0ire commented:
Ok, this is looking like it should :)

Kertudex commented:
It should be OK now. 
I added a line for the unit test instead of removing the one asked by @OskarStark, removing it is breaking the test. 
The default value is in any case pushing an 'actions: [list]' value. 
Again, thank you very much for the tutorial, rebase is understood for me now :)

greg0ire commented:
> thank you very much for the tutorial, rebase is understood for me now 

Heh. There is one tiny thing to fix apparently : 

> 1) Tests/DependencyInjection/ConfigurationTest.php (align_double_arrow)

greg0ire commented on Tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -32,12 +32,14 @@ public function testProcessing()
                 array('admin_finder_services' => array(
                     'my_admin' => array(
                         'finder' => 42,
```
I think this is the line that is faulty.

Kertudex commented:
I stashed the last commits. Build is ok now.

greg0ire commented:
It is indeed! Merging!

Kertudex commented:
Thank you !

greg0ire commented:
No, thank _you_! Have a good week!

OskarStark commented:
Thank you very much

Am Montag, 7. Dezember 2015 schrieb Grégoire Paris :

> No, thank _you_! Have a good week!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/sonata-project/SonataAdminSearchBundle/pull/22#issuecomment-162361276
> .

Overdeath commented:
Hi,
Is it possible that this change broke the global search feature?
When i use the global search the following error message is displayed "The Request object has not been set". I think this happens because in the global search context the request is not accesible so "$admin->getRequest()" probably throws an error. 
I managed to get around this and have it working by adding a check for $admin->hasRequest(). Here is what i did https://github.com/sonata-project/SonataAdminSearchBundle/commit/ccc86eeb02654ceb83294e6ba68c1aacf75620a8 

This has some drawbacks because you no longer have control over where elastic is used. 
Is there maybe another way to get the action?

greg0ire commented:
I think the best way would be to inject the request stack service via the constructor of the builder.
